examples/step-64/step-64.cc

```diff
@@ -0,0 +1,386 @@
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/grid/manifold_lib.h>
+
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping_q_cache.h>
+#include <deal.II/fe/mapping_q_generic.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/vector_tools.h>
+
+
+
+namespace step64
+{
+  using namespace dealii;
+
+  template <int dim>
+  class ExactSolution : public Function<dim>
+  {
+  public:
+    virtual double value(const Point<dim> &p,
+                         const unsigned int /*component*/ = 0) const override
+    {
+      if (p.norm_square() < 0.25)
+        return p.norm_square();
+      else
+        return 0.1 * p.norm_square() + (0.25 - 0.025);
+    }
+
+    virtual Tensor<1, dim>
+    gradient(const Point<dim> &p,
+             const unsigned int /*component*/ = 0) const override
+    {
+      if (p.norm_square() < 0.25)
+        return 2. * p;
+      else
+        return 0.2 * p;
+    }
+  };
+
+
+  template <int dim>
+  class Coefficient : public Function<dim>
+  {
+  public:
+    virtual double value(const Point<dim> &p,
+                         const unsigned int /*component*/ = 0) const override
+    {
+      if (p.norm_square() < 0.25)
+        return 0.5;
+      else
+        return 5.;
+    }
+  };
+
+
+
+  template <int dim>
+  class Step64
+  {
+  public:
+    Step64();
+    void run();
+
+  private:
+    void create_grid();
+    void setup_system(const Mapping<dim> &mapping);
+    void assemble_system(const Mapping<dim> &mapping);
+    void solve();
+    void output_results(const Mapping<dim> &mapping) const;
+
+    Triangulation<dim> triangulation;
+    FE_Q<dim>          fe;
+    DoFHandler<dim>    dof_handler;
+
+    AffineConstraints<double> constraints;
+    SparsityPattern           sparsity_pattern;
+    SparseMatrix<double>      system_matrix;
+    Vector<double>            solution;
+    Vector<double>            system_rhs;
+
+    mutable TimerOutput timer;
+  };
+
+
+
+  template <int dim>
+  Step64<dim>::Step64()
+    : fe(3)
+    , dof_handler(triangulation)
+    , timer(std::cout, TimerOutput::never, TimerOutput::wall_times)
+  {}
+
+
+
+  template <int dim>
+  void Step64<dim>::create_grid()
+  {
+    Triangulation<dim> tria_outer, tria_inner;
+    GridGenerator::hyper_shell(
+      tria_outer, Point<dim>(), 0.5, std::sqrt(dim), 2 * dim);
+    GridGenerator::hyper_ball(tria_inner, Point<dim>(), 0.5);
+    GridGenerator::merge_triangulations(tria_inner, tria_outer, triangulation);
+    triangulation.reset_all_manifolds();
+    triangulation.set_all_manifold_ids(0);
+    for (const auto &cell : triangulation.cell_iterators())
```

Comment:
After the release is out I should find some time to finish #7052 so that we do not need to write this kind of code anymore.